File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.py
import heapq

class Solution:
    def furthestBuilding(self, heights: List[int], bricks: int, ladders: int) -> int:
        heap = []
        answer = len(heights) -1
        for i in range(1, len(heights)):
            if heights[i] - heights[i-1] > 0:
                print(heights[i] - heights[i-1])
                heapq.heappush(heap, heights[i] - heights[i-1])
            
            if len(heap) > ladders:
                used = heapq.heappop(heap)
                if bricks - used < 0:
                    answer = i -1
                    break
                bricks -= used
                    
        return answer
            

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.kt
class Solution {
    fun furthestBuilding(heights: IntArray, bricks: Int, ladders: Int): Int {
        val pq = PriorityQueue<Int>()       
        var answer = heights.size -1
        var totalBricks = bricks
        for (i in 1 until heights.size){
            if (heights[i] -heights[i-1] > 0){
                pq.add(heights[i] -heights[i-1])
            }
            
            if (pq.size > ladders){
                val usedBrick = pq.poll()
                if (totalBricks - usedBrick < 0){
                    answer = i - 1
                    break
                }
                totalBricks -= usedBrick
            }
        }

        return answer
    }
}